txpipe/psf_diagnostics.py

```diff
@@ -272,9 +273,373 @@ def save_stats(self, rowe_stats):
+                plt.xlabel(r"$\theta$")
+                plt.ylabel(r"$\xi_+(\theta)$")
+                plt.legend()
+        f.close()
+
+    def save_stats(self, galaxy_star_stats):
 
+        f = self.open_output('star_shear_stats')
+        g = f.create_group("star_cross_galaxy")
+        #TODO add a more informative name than 1,2 for this type
+        for i in 1,2,3,4:
+            for s in STAR_TYPES:
+                theta, xi, err = galaxy_star_stats[i, s]
+                name = STAR_TYPE_NAMES[s]
+                h = g.create_group(f'star_cross_galaxy_{i}_{name}')
+                h.create_dataset('theta', data=theta)
+                h.create_dataset('xi_plus', data=xi)
+                h.create_dataset('xi_err', data=err)
+        f.close()
 
 
+
+
+class TXStarDensityTests(PipelineStage):
+    """
+    Star cross galaxy and star autocorrelation density stats.
+    """
+
+    name = 'TXGalaxyStarDensity'
+
+    inputs =[('shear_catalog', ShearCatalog),('star_catalog', HDFFile),
+    ('shear_tomography_catalog', TomographyCatalog),
+    ('random_cats', RandomsCatalog)]
+    outputs = [
+        ('star_density_test', PNGFile),
+        ('star_density_stats', HDFFile),
+    ]
+
+    config_options = {
+        'min_sep':0.5,
+        'max_sep':250.0,
+        'nbins':20,
+        'bin_slop':0.1,
+        'sep_units':'arcmin',
+        'psf_size_units':'sigma',
+        'shear_catalog_type':'metacal',
+        'flip_g2': False,
+    }
+
+    def run(self):
+        import treecorr
+        import h5py
+        import matplotlib
+        matplotlib.use('agg')
+
+        ra, dec,  star_type = self.load_stars()
+        ra_gal, dec_gal = self.load_galaxies()
+        ra_random, dec_random = self.load_randoms()
+
+        #only use reserved stars for this statistics
+        galaxy_star_stats = {}
+        for t in STAR_TYPES:
+            s = star_type == t
+            galaxy_star_stats[1, t] = self.compute_galaxy_star(1, ra, dec, s, ra_gal, dec_gal, ra_random, dec_random)
+            galaxy_star_stats[2, t] = self.compute_star_star(2, ra, dec, s, ra_gal, dec_gal, ra_random, dec_random)
+
+        self.save_stats(galaxy_star_stats)
+        self.galaxy_star_plots(galaxy_star_stats)
+
+
+
+    def load_stars(self):
+        f = self.open_input('star_catalog')
+        g = f['stars']
+        ra = g['ra'][:]
+        dec = g['dec'][:]
+
+        star_type = load_star_type(g)
+
+        return ra, dec, star_type
+
+    def load_randoms(self):
+
+        f = self.open_input('random_cats')
+        group = f['randoms']
+        ra_random = group['ra'][:]
+        dec_random = group['dec'][:]
+        return ra_random, dec_random
+
+    def load_galaxies(self):
+
+        # Columns we need from the shear catalog
+        # TODO: not sure of an application where we would want to use true shear but can be added
+        read_shear_catalog_type(self)
+
+        # load tomography data
+        f = self.open_input('shear_tomography_catalog')
```

Comment:
Doesn't matter here but in case useful later, you can now do
```
with self.open_input('shear_tomography_catalog') as f:
    ... #do stuff with f
```
and it closes automatically.